### PR TITLE
Payola supports Rails 5.0

### PR DESCRIPTION
**Pull request by eliotsykes**
- [x] Run test suite against Rails 4.1, 4.2, 5
- [x] Fix Rails 5 deprecation warnings
- [x] Regenerate spec/dummy app with Rails 5 to check for any significant differences to incorporate
- [x] Rebase
- [x] Squash commits

### CODE REVIEW

**Comment by deanpcmad** [comment]
👍

**Comment by eliotsykes** [comment]
📰 **Good news!** The CI environment is running the tests against Rails 5.0, 4.2, 4.1!
🗞 **Bad news** The specs aren't passing for Rails 4.2, 4.1.
📰 **Good news again!** The spec failures look as expected and I intend to deal with them shortly 🤓

**Comment by eliotsykes** [comment]
TIL: Sqlite 3.7.9 does not play nicely with Rails 5.0 migrations. The schema_migrations table will only contain the very first and very last migration version. Upgrading Sqlite to 3.8.2 on CircleCI fixes this. I've an unproven suspicion this is related to what was encountered in rails/rails#24440

File: spec/support/controller_spec_compatibility.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  #     super(path, params: params, headers: headers)
+  #   end
+  # end
+
+  def permitted_params(params)
+    ActionController::Parameters.new(params).permit!
+  end
+end
+
+RSpec.configure do |config|
```
Comment (peterkeen) [inline spec/support/controller_spec_compatibility.rb]: Doesn't this break strong params? Is this just a temporary testing shim?

**Comment by peterkeen** [comment]
This is great work @eliotsykes ! Thanks!

When you're ready to merge (no rush or anything) we should talk about squashing the WIP commits down into a readable changeset.

File: spec/support/controller_spec_compatibility.rb (hunk 1 above)
Comment (eliotsykes) [inline spec/support/controller_spec_compatibility.rb]: > Is this just a temporary testing shim?

Yep! Needs work, its hopefully going to be a way to stop Rails 5.0 deprecation notices for `get`, `post` etc. calls in controller specs that don't use keyword args.

**Comment by eliotsykes** [comment]
You're welcome @peterkeen - thanks for writing such a valuable gem!

Regarding the squashing when the time comes, I'm happy to squash into a single commit for this branch or break it up into a few commits, whatever sits best with you.

**Comment by eliotsykes** [comment]
Specs now passing on all supported Rails versions. Next up is getting rid of deprecation warnings from the Rails 5 suite:

```
DEPRECATION WARNING: `:nothing` option is deprecated and will be removed
in Rails 5.1. Use `head` method to respond with empty response body.
(called from render_payola_status at 
/home/ubuntu/payola/app/controllers/concerns/payola/status_behavior.rb:6)

DEPRECATION WARNING: `redirect_to :back` is deprecated and will be removed 
from Rails 5.1. Please use `redirect_back(fallback_location: fallback_location)` 
where `fallback_location` represents the location to use if the request has no 
HTTP referer information. 
(called from update at 
/home/ubuntu/payola/app/controllers/payola/customers_controller.rb:9)

DEPRECATION WARNING: Method to_hash is deprecated and will be removed 
in Rails 5.1, as `ActionController::Parameters` no longer inherits from hash. 
Using this deprecated behavior exposes potential security problems. If you 
continue to use this method you may be creating a security vulnerability in 
your app that can be exploited. Instead, consider using one of these 
documented methods which are not deprecated: 
http://api.rubyonrails.org/v5.0.0/classes/ActionController/Parameters.html 
(called from merge at 
/home/ubuntu/payola/vendor/bundle/ruby/2.3.0/gems/stripe-1.20.1/lib/stripe/api_operations/update.rb:5)

DEPRECATION WARNING: `:nothing` option is deprecated and will be removed 
in Rails 5.1. Use `head` method to respond with empty response body. 
(called from render_payola_status at 
/home/ubuntu/payola/app/controllers/concerns/payola/status_behavior.rb:6)

DEPRECATION WARNING: `redirect_to :back` is deprecated and will be removed 
from Rails 5.1. Please use `redirect_back(fallback_location: fallback_location)` 
where `fallback_location` represents the location to use if the request has no 
HTTP referer information. 
(called from destroy at 
/home/ubuntu/payola/app/controllers/payola/cards_controller.rb:18)

DEPRECATION WARNING: `redirect_to :back` is deprecated and will be removed 
from Rails 5.1. Please use `redirect_back(fallback_location: fallback_location)` 
where `fallback_location` represents the location to use if the request has no 
HTTP referer information. 
(called from create at 
/home/ubuntu/payola/app/controllers/payola/cards_controller.rb:9)

DEPRECATION WARNING: `redirect_to :back` is deprecated and will be removed 
from Rails 5.1. Please use `redirect_back(fallback_location: fallback_location)` 
where `fallback_location` represents the location to use if the request has no 
HTTP referer information. 
(called from create at 
/home/ubuntu/payola/app/controllers/payola/cards_controller.rb:11)
```

**Comment by jordanyeo** [comment]
@eliotsykes I've been using your fork on a new Rails 5 project, and when trying to execute the Cancel Subscription functionality, I'm getting an error. Tracked it down, and it looks like you need to change Line 24 of SubscriptionsController.rb to 

`Payola::CancelSubscription.call(subscription, at_period_end: ActiveRecord::Type::Boolean.new.cast(params[:at_period_end]))`

**Comment by eliotsykes** [comment]
@jordanyeo Thanks so much for the feedback 👍. 

This might be fixed by updating to the latest revision of the `rails-5.0` branch on the fork.  

After `bundle update payola-payments` see if the bug continues. If it does, check the bundled `Payola::SubscriptionsController` contains the changes in commit https://github.com/peterkeen/payola/pull/232/commits/ea039a3ef61a145027010212c727a11e95eb17eb and please report back anything you find out.

**Comment by jordanyeo** [comment]
Whoops! Wasn't up to date. Thanks so much! :)

**Comment by eliotsykes** [comment]
Progress update: Work continues on task "Regenerate spec/dummy app with Rails 5 to check for any significant differences to incorporate". 

I'm going through the diff one file at a time and am about a third of the way through this task. Will continue throughout the week.

**Comment by jordanyeo** [comment]
@eliotsykes again, thank you SO much for your work. Just ran into another issue, wasn't sure if it's on your radar. When traversing down to VERSION=0 with database migrations in my rails 5 app, I get this error:

`== 20151205004838 ChangeTaxPercentFormatInPayolaSubscriptions: reverting ======
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:



This migration uses change_column, which is not automatically reversible.
To make the migration reversible you can either:
1. Define #up and #down methods in place of the #change method.
2. Use the #reversible method to define reversible behavior.`

I wonder if it would even be possible to consolidate the migrations down to a fewer set- but I guess that would probably break the ability to upgrade. For fresh installs, might be nicer to rely on a minimal set of migrations.

Thanks again!

**Comment by eliotsykes** [comment]
@jordanyeo Thanks for the thanks and for the migration feedback. I see the appeal of fewer migrations too for fresh installs though I think you're right about it impacting devs negatively who want to upgrade existing payola installations.

**Comment by jordanyeo** [comment]
@eliotsykes  Want me to do a pull request with the migration edited to have up and down methods?

**Comment by eliotsykes** [comment]
@jordanyeo I'm only a contributor here

**Comment by peterkeen** [comment]
@jordanyeo I'd love to see a PR that fixes that migration.

**Comment by eliotsykes** [comment]
@peterkeen I think, finally, this is about done 😀  I'd planned on squashing the commits into one with a detailed message explaining the significant changes. Any feedback from everyone more than welcome!

**Comment by jordanyeo** [comment]
Working great for me! I'd so go ahead and merge! Again, thanks so much for your work on this!

**Comment by peterkeen** [comment]
@eliotsykes Thanks! 👍 I would prefer a squash of one or a small number of commits, but it's up to you.

**Pull request merged**